Advanced Shield Generators

atlantiscze opened the pull request:
Creates new advanced shield generators, designed to replace old hull and bubble shield generators. They address various problems, as well as add new possibilities. This PR also involves some meteor related changes and changes to the Meteor game mode which is now votable and quite suitable for normal gameplay.
- This is a squashed commit that adds a new module shield_generators.
- Axes the renwick concept. Instead, shields are purely based on the amount of energy you can push into the shield. Shield generators are built with a SMES coil, and use it to store energy. Damage to the field uses up this energy, as well as upkeep.
- Shield generator on it's own has only little upkeep, but is useless. The proper use is given by shield modes, which can be enabled in the generator's UI. Each shield mode tweaks the field by adding a new functionality (for example: Hyperkinetic mode blocks bullets, meteors and various other objects).
- Field upkeep increases exponentially on a per-mode basis. Each mode has a set multiplier. These multipliers affect the upkeep value. That means fields with only one or two modes are much less energy intensive than fields with all modes enabled.
- Field no longer fully blocks carps. Carps (as well as other mobs) will now try to bite their way through the field, which can be quite severe threat for the field.
- If the generator's stored energy is lower than 35% there is a risk of temporary field breach when the field takes damage. This probability grows as the stored energy drops. If a hit uses up all remaining energy the field collapses completely and goes into recovery mode. Once stored energy gets above 25% the field restarts again. Alternatively, it may be restarted manually before this point.
- Another noteworthy feature is shield mitigation. Shields can go up to 60 - 80% (basic capacitor vs super capacitor used in construction) mitigation for three damage types (EM/Physical/Heat). Basically this means the field adapts to the taken damage. If you keep hitting the field with a blunt object, physical mitigation will grow, while EM and heat will drop, making the field more resistant to physical damage.
- Meteors have been tweaked a bit. Tunguska is still very dangerous, but to a lower degree (a meteor piercing through a whole station made of steel was silly and way too OP). They also no longer pierce through shields, assuming the hyperkinetic mode is enabled. They can, however, put a giant stress on the shield.
- Shield Diffusers have been added. There are two variants - floor mounted variant, that is APC powered and can be built from few components. It layers under floor tiles. Second variant is handheld, and it is powered by an internal cell. This one replaces the shield disrupter. traitor item which is no longer useful.
- Floor mounted shield diffusers automatically turn off during meteor storm events. Meteor storms first give a warning announcement. After one minute diffusers turn off (indicated by a blinking red-yellow light on the diffuser). After additional 20~ seconds the meteors begin as usual.
- An attempt has been made to revive the Meteor gamemode. It is now votable, and less disastrous. It has 2x increased rate of random events, as well as increasingly frequent meteor strikes after 30 minute mark. Meteors will start as few tiny dust impacts, but will increase in severity over time. Two warning announcements will be sent before this happens. Shields are probably the best way for extended survival of this mode. Hopefully this will make the mode an interesting and more PVE focused (rather than antag focused) alternative to Extended.

TODO list:
- [x] Remove old shield generators as they are fully obsolete now.
- [x] Mapping changes
- [x] Changelog

atlantiscze committed:
Initial commit (Squashed)

atlantiscze committed:
Removes old shield generators
- NOTE: This commit will fail travis, as it does not involve map changes. Map changes will be done in a single next commit.

atlantiscze committed:
Map changes
- Removes old shield generators which have been removed from code in previous commit.
- Adjusts exterior areas, for the purpose of better calibration of hull shield.
- Scrubbers in xenobiology and virology are no longer on special pipelines. The option to vent them into space was never used anyway, and they remained bridged into main scrubber pipe. Therefore the vents to space could be axed too (in these two areas)
- Maps in floor mounted shield diffusers to various airlocks, shuttle docks, etc.

atlantiscze committed:
Changelog Entry

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,300 @@
+
+
+// As we have various shield modes, this handles whether specific things can pass or not.
+/obj/effect/shield/CanPass(var/atom/movable/mover, var/turf/target, var/height=0, var/air_group=0)
```
Make an /atom/movable/can_pass_shield_proc() and override appropriately instead is these istype checks.

Techhead0 commented on code/modules/shield_generators/shield_generator.dm:
```diff
@@ -0,0 +1,521 @@
+		"status" = check_flag(MODEFLAG_HYPERKINETIC),
+		"hacked" = 0,
+		"multiplier" = MODEUSAGE_HYPERKINETIC
+		)))
```
Instead of flags and these odd lists and editing stuff in defines, you should probably just have the modes as datums. It feels like it would be much cleaner this way.

atlantiscze replied:
I used bitflags because it's easy to manipulate them. This proc mostly serves to create an user-friendly format for the UI. Defines are used because it allows for very easy tweaking, simply by adjusting one file, without having to dig around multiple files changing variables directly.

PsiOmegaDelta replied:
You call the long list of list manipulation below more maintainable than keeping mode datums?

atlantiscze replied:
As i have already said, this proc only serves as a source of data for NanoUI. That's why all the lists are there - NanoUI requires that format if i want to pass all information about shield modes in a single data[] entry.

I feel that actual mode datums are unnecessary at this point. Modes themselves don't do anything specific - all the work is done by shield generators, which only check the bitflags to determine which modes are enabled and which are not. Having to manipulate lists of enabled/disabled modes sounds more complicated to me than having simple bitflags for this simple purpose.

PsiOmegaDelta replied:
Well, you're free to make this flag info creation more generic in any way you want. As is it's not considered maintainable.

caelaislinn replied:
Store the lists outside the proc (in the parent atom), have ui_interact pass those saved lists every time time its called, then update the lists elsewhere as needed.

caelaislinn replied:
This many bitflags shouldn't be checked individually like this as you're passing a lot of unnecessary data and structures. A better solution would be to simply pass the defines into the NanoUI template, then have a switch for the strings in there. 

From what I'm aware NanoUI doesn't allow datum passing (?) so we're limited to multidimensional or associative lists for passing more complex data structures.

caelaislinn replied:
Aka nanui may be lacking functionality to make this any more "maintainable"

PsiOmegaDelta replied:
I'll just open a PR which changes to this PR myself to show you how it's done.

PsiOmegaDelta commented:
What are the map changes about? In particular Exodus cargo maintenance that now appears to be exposed to space.

PsiOmegaDelta commented on code/modules/shield_generators/floor_diffuser.dm:
```diff
@@ -0,0 +1,69 @@
+
+/obj/machinery/shield_diffuser/attack_hand()
+	if(alarm)
+		usr << "You press an override button on \the [src], re-enabling it."
```
`to_chat(usr, message)` 
Repeat for all instances of `X << Text`

atlantiscze replied:
I have no problem with using this, but for clarity, what is the difference between << and to_chat()? I tried searching for "proc/to_chat(" but it's not there, so i assume it is some kind of BYOND method?

Techhead0 replied:
`to_chat(x,y)` is a define that outputs `x << y`. Someday it may be replaced with a fancy proc.

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,300 @@
+	icon = 'icons/obj/machines/shielding.dmi'
+	icon_state = "shield_normal"
+	anchored = 1
+	layer = 4.1		// Above mobs
```
```
plane = ABOVE_HUMAN_PLANE
layer = ABOVE_HUMAN_LAYER // Or otherwise adjusted with a relevant constant in _planers+layers.dm
```

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,300 @@
+
+
+/obj/effect/shield/Destroy()
+	..()
```
`. = ..()`
Repeat for all other Destroy calls where the result of ..() isn't returned.

atlantiscze commented:
The map changes are 95% related to exterior of the station. Hull shield mode acquires tiles in slightly different way.

Currently, the method is as follows. It goes through all tiles in range, and detetermines if it's a valid tile. Valid tile is a non-space and non-shuttle tile, both in terms of turfs and areas. In other words, various junk, antagonist shuttles (this was the primary intent) and random floor tiles that don't have an area set will be ignored.

As shuttles are no longer protected, i added few protective panels with grilles and reinforced walls to reduce the risk of shuttles being hit (mostly for the purpose of the Meteor mode so shuttles aren't guaranteed to be destroyed). These grilles had floor tiles under them, with an area set to nearest maintenance. That's what caused the travis build failure, so i'll probably resolve it in a different way.

edit: The remaining 5% is removal of old mapped in shield generators (i replaced them with bunch of canisters because the empty space looked weird), and addition of some circuits to tech storage so the new shields and diffusers can be built.

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,300 @@
+			return ..()
+		var/obj/effect/meteor/M = mover
+		take_damage(M.get_shield_damage(), SHIELD_DAMTYPE_PHYSICAL, M)
+		visible_message("<span class='danger'>\The [M] breaks nto dust!</span>")
```
nto dust.

PsiOmegaDelta commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -127,7 +128,8 @@
 /obj/effect/meteor/Bump(atom/A)
-	if(A)
+	..()
+	if(A && !gcDestroyed)	// Prevents explosions and other effects when we were deleted by whatever we Bumped() - currently used by shields.
```
`!deleted(src)`

PsiOmegaDelta commented on code/modules/shield_generators/shield_generator.dm:
```diff
@@ -0,0 +1,521 @@
+		return
+	for (var/x_offset = -field_radius; x_offset <= field_radius; x_offset++)
+		for (var/y_offset = -field_radius; y_offset <= field_radius; y_offset++)
+			T = locate(gen_turf.x + x_offset, gen_turf.y + y_offset, gen_turf.z)
```
This seems to re-implement /trange() with the exception of allowing rectangular rather than strictly square shapes, same with above.

PsiOmegaDelta commented on code/modules/shield_generators/shield_generator.dm:
```diff
@@ -0,0 +1,521 @@
+		. = 1
+
+	if(.)
+		nanomanager.update_uis(src)
```
You don't need to manually call update_uis() for src in Topic() if 1 is returned appropriately.

PsiOmegaDelta commented on code/modules/shield_generators/shield_generator.dm:
```diff
@@ -0,0 +1,521 @@
+		return 1
+
+	if(issilicon(user) && !Adjacent(user) && ai_control_disabled)
+		return
```
Consider moving this check to `/CanUseTopic(var/mob/user)` instead.

PsiOmegaDelta commented on code/modules/shield_generators/handheld_diffuser.dm:
```diff
@@ -0,0 +1,45 @@
+
+/obj/item/weapon/shield_diffuser/New()
+	cell = new(src)
+	processing_objects.Add(src)
```
Should preferably only add objects to the processing queue if they actually need it, i.e. add/remove in /attack_self(), remove in Destroy() only if currently active.

PsiOmegaDelta commented on code/modules/shield_generators/handheld_diffuser.dm:
```diff
@@ -0,0 +1,45 @@
+	..()
+
+/obj/item/weapon/shield_diffuser/Destroy()
+	qdel(cell)
```
cell = null

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,300 @@
+	diffused_for = max(duration, 0)
+	gen.damaged_segments |= src
+	density = 0
+	invisibility = 101
```
INVISIBILITY_MAXIMUM

PsiOmegaDelta commented:
Meteor mode is pretty weak if shuttles remain functional. People have already shown that they're more than willing to shuttle off the station and wait the round out.

atlantiscze commented:
They remain functional, but with a risk. Shuttles have varying levels of protection, with Departures/Arrivals/Escape Pods being a bit safer than the others. The intention of these reinforcement panels is to provide grounds for hull shield to form on, so if they are set up the most used shuttles won't be affected that much. Without shield, a flimsy grille with an occassional wall won't block more than one meteor anyway. It can always be removed, but i didn't want to completely lock out people from the possibility to transfer to the outposts and back to do their jobs. The whole idea behind new meteor mode is to have it as a PVE oriented variant of extended, however thanks to the shields it can be quite survivable, with some luck (though a grid check combined with unlucky timing may still be very fatal). It also has double amount of random events, which should hopefully make it a "less calm extended".

atlantiscze committed:
Addresses Github comments.

PsiOmegaDelta commented on code/modules/shield_generators/shield_generator.dm:
```diff
@@ -250,14 +250,15 @@
 
+/obj/machinery/power/shield_generator/CanUseTopic(var/mob/user)
+	if(issilicon(user) && !Adjacent(user) && ai_control_disabled)
+		return 0
```
STATUS_CLOSE or STATUS_UPDATE, whichever you deem most appropriate.
0 is equal to STATUS_DISABLED which means the windows open but never refreshes/updates.

Raptor1628 commented:
I don't like nerfing the Tunguska, I've only seen it twice when not bussed in over the entire time of play. I'm fine with it not putting a hole in shields, just damaging them, but it should be the major threat that it is considering how rare it is.

atlantiscze commented:
It is still about two times stronger than the 2nd strongest meteor.

I found it silly that a single meteor can pierce through whole station. That's both highly unrealistic, and also bad for gameplay as fixing the damage it causes is nearly impossible when the hole spans from one side of the station to the other. After this change it's still dangerous, but not to such massive degree. As for tunguska and shields, Tunguska causes about 50MJ of damage to the shield generator, which is more than half of a regular generator's capacity.

caelaislinn commented:
Renwicks were just an arbitrary ratio to allow balancing of energy <-> shield conversion. There were also mechanics built in (stress and the charging thing) to prevent simply hardwiring the entire station energy output into the shields to make them nearly unkillable and thus really unbalancable.

atlantiscze commented:
I've intentionally replaced the renwick system with actual energy. I feel that people should be rewarded if they can provide enough power to the shield, not punished. It requires some effort, either in the form of setting up a buffer SMES or by increasing power output of available power sources.

The power usage is balanced in a way where 1MW is, more or less, enough to have a rudimentary anti-meteor and anti-carp shield. If you want extra functions, such as atmosphere containment, you have to proportionally increase the power input. The power usage increases exponentially with each enabled mode. A shield with all modes enabled is unrealistic on a stationwide scale given the tremendeous upkeep in tens of megawatts. Though, it may be more feasible on a smaller scale.

caelaislinn commented:
Yeah one of my sidegoals was to provide a high end power gobbler for the station, but I tried to balance it so that it wouldn't be something you just dump all power ever into. The exponential increase for each feature is a cool idea. 

I personally would have made it machine component based so that engineers have to manually construct the generator with the right components for the effect they want, instead of having a (literal) deus ex machina that can do everything if you press the right buttons.

PsiOmegaDelta commented:
Diff too big to view. Merge conflicts.

caelaislinn commented:
It looks like a bunch of unrelated changes got in where a bunch of magic numbers were changed to defines. Good idea, wrong PR. Maybe it was a muckup with updating mid-WIP?

atlantiscze commented:
It seems that my last dev merge didn't go exactly as planned. I'll fix that on Sunday, as i'm currently limited to a combination of very bad computer and internet connection.

atlantiscze commented:
I've removed the last commit that caused most of the trouble, but i'm having problems updating the branch.

I use the `git pull bay dev` command which usually does the trick, but this time it seems to fail due to torch once again.

When the automatically run mapmerge encounters torch-10, the following message is printed:
![image](https://cloud.githubusercontent.com/assets/5147960/19725420/b14df2be-9b85-11e6-9eb8-769686bfea47.png)

And, at the end, when other maps/files are successfully merged the following error occurs:
![image](https://cloud.githubusercontent.com/assets/5147960/19725430/c11ec2ea-9b85-11e6-8e7d-b9cfdea5cca3.png)

I am genuinely lost once again. I've found a page that suggested running `git fsck --full` - which didn't seem to help with anything. Any ideas?

PsiOmegaDelta commented:
Far as I can tell from a quick search, add_info cached failed basically means you might as well start over from the 'git clone' step.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-10-13-Shields

atlantiscze commented:
I have no idea what caused the problem, but it seems to be gone now.

PsiOmegaDelta commented:
Only now you have a 5000 line diff.

atlantiscze commented:
It seems torch-10.dmm is the culprit. I'll resolve that one manually.

recharger.dm seems to also have a certain oddity - github considers the whole file changed, but i've only touched few lines. I recall seeing a discussion about line endings few days ago (not sure whether it was on IRC or in another PR), can that cause this kind of problem?

PsiOmegaDelta commented:
See if a \n -> \r\n application in the affected file will sort it out.

atlantiscze committed:
Additional minor tweaks
- Shield modes are now partially datum-based
- Also should fix a runtime problem with atmospheric containment.

atlantiscze commented:
I've had to checkout the dev version of the file, and re-apply my changes to it. The problem was probably due to #14480 as my branch was created prior to that PR.

I've also tried to improve the unpopular proc that generated NanoUI data. It is now partially replaced with datums that control the power usage multiplier, and contain the definition of names/descriptions.

Either way, it should be all good for merge now.

PsiOmegaDelta commented:
That PR did indeed correct all newlines to the appropriate format. I don't expect that to be undone.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-10-13-Shields

PsiOmegaDelta commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -97,26 +97,29 @@
 	. = ..() //process movement...
 
-	if(.)//.. if did move, ram the turf we get in
+	if(. && !istype(get_turf(src), /turf/space/))
```
This basically means a meteor won't ex_act() non-dense objects on space turfs.

atlantiscze replied:
This is unfortunately necessary due to how meteors and explosions work. It is part of a solution that resolves a problem where meteor exploded right next to the shield, and explosion passed through the shield and caused damage.

In general, if you have any suggestions on how to block explosions, i'll gladly welcome them.

PsiOmegaDelta replied:
We apparently have recursive explosions since #13873 (or the option to enable them). It should be possible to make them respect shields.

PsiOmegaDelta replied:
Also, meteors still have the possibility of hitting a dense object near shields and explode through them.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-10-13-Shields

atlantiscze committed:
Now supports recursive explosions
- They are blocked by the Hyperkinetic Projectiles flag.

PsiOmegaDelta commented on code/game/gamemodes/meteor/meteors.dm:
```diff
@@ -116,7 +116,7 @@
+	if(.)
 		var/turf/T = get_turf(loc)
 		ram_turf(T)
 		if(prob(20))
```
If you don't move the /turf/space check back to this line meteors have a 20% chance to somehow get hit by nothing.

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,314 @@
+		M.make_debris()
+		qdel(M)
+		return 0
+	if(istype(mover, /mob/living/))
```
Basically the same thing as with can_pass_shield. Proc overrides before istype checks.

PsiOmegaDelta commented on code/modules/power/power.dm:
```diff
@@ -15,6 +15,10 @@
 	idle_power_usage = 0
 	active_power_usage = 0
 
+/obj/machinery/power/New()
```
Potentially best left to /initialize()

PsiOmegaDelta commented on code/modules/shield_generators/shield.dm:
```diff
@@ -0,0 +1,318 @@
+/obj/effect/shield/New()
+	..()
+	update_nearby_tiles()
+	s = new /datum/effect/effect/system/spark_spread(src)
```
Doesn't appear to be cleaned up during destroy.

atlantiscze replied:
Resolved, as well as bump changes and everything else mentioned.

PsiOmegaDelta commented on code/modules/shield_generators/shield_generator.dm:
```diff
@@ -0,0 +1,434 @@
+	field_segments = null
+	damaged_segments = null
+	mode_list = null
+	qdel(wires)
```
wires var isn't nulled. There is a handy qdel_null() macro for that now tho.

PsiOmegaDelta reviewed (changes requested):
Bump handling aside I suppose only the few mentioned issues remain.

PsiOmegaDelta commented:
Bump handling is still what it is but I suppose it's otherwise OK.
